### PR TITLE
switch to kube authorization attributes

### PR DESCRIPTION
**Pull request by deads2k**
Builds on https://github.com/openshift/origin/pull/13259.

This pull removes our existing attributes and uses the upstream kube attributes.  At this point our interfaces should overlap.  I'm planning one more to remove our authorizer interface altogether.

@liggitt @enj

### CODE REVIEW

**Comment by deads2k** [comment]
[test]

File: pkg/authorization/authorizer/attributes_builder.go [hunk 1]
```diff
@@ -16,28 +17,36 @@ func NewAuthorizationAttributeBuilder(contextMapper kapi.RequestContextMapper, i
 	return &openshiftAuthorizationAttributeBuilder{contextMapper, infoFactory}
 }
 
-func (a *openshiftAuthorizationAttributeBuilder) GetAttributes(req *http.Request) (Action, error) {
+func (a *openshiftAuthorizationAttributeBuilder) GetAttributes(req *http.Request) (authorizer.Attributes, error) {
+	attribs := authorizer.AttributesRecord{}
```
Comment (ncdc) [inline pkg/authorization/authorizer/attributes_builder.go]: Move this down before line 32?

File: pkg/authorization/authorizer/cache/authorizer.go [hunk 2]
```diff
@@ -122,22 +122,22 @@ func (c *CacheAuthorizer) GetAllowedSubjects(ctx kapi.Context, attributes author
...
+		"readOnly":        a.IsReadOnly(),
+		"resourceRequest": a.IsResourceRequest(),
+		"path":            a.GetPath(),
 	}
 
-	if namespace, ok := kapi.NamespaceFrom(ctx); ok {
-		keyData["namespace"] = namespace
-	}
-	if user, ok := kapi.UserFrom(ctx); ok {
+	keyData["namespace"] = a.GetNamespace()
```
Comment (ncdc) [inline pkg/authorization/authorizer/cache/authorizer.go]: Move up into the map declaration?

File: pkg/authorization/authorizer/cache/authorizer_test.go [hunk 3]
```diff
@@ -20,35 +18,34 @@ func TestAuthorizer(t *testing.T) {
...
+			Attrs: authorizer.AttributesRecord{
+				User:            &user.DefaultInfo{Name: "me", Groups: []string{"group1", "group2"}},
+				Verb:            "v",
+				Namespace:       "myns",
+				APIVersion:      "av",
+				APIGroup:        "ag",
+				Resource:        "r",
+				Subresource:     "sub",
+				Name:            "rn",
+				ResourceRequest: true,
```
Comment (liggitt) [inline pkg/authorization/authorizer/cache/authorizer_test.go]: doesn't really matter, but `NonResourceURL: true` is the opposite of `ResourceRequest: true`

File: pkg/authorization/authorizer/cache/authorizer_test.go [hunk 4]
```diff
@@ -59,7 +56,7 @@ func TestCacheKey(t *testing.T) {
 }
 
 func TestCacheKeyFields(t *testing.T) {
-	keyJSON, err := cacheKey(kapi.NewContext(), &authorizer.DefaultAuthorizationAttributes{})
+	keyJSON, err := cacheKey(authorizer.AttributesRecord{User: &user.DefaultInfo{Name: "me"}})
```
Comment (liggitt) [inline pkg/authorization/authorizer/cache/authorizer_test.go]: why was the default user needed?

Comment (liggitt) [inline pkg/authorization/authorizer/cache/authorizer_test.go]: oh, to exercise all the cache keys

File: pkg/authorization/authorizer/cache/authorizer_test.go (hunk 3 above)
Comment (deads2k) [inline pkg/authorization/authorizer/cache/authorizer_test.go]: > doesn't really matter, but NonResourceURL: true is the opposite of ResourceRequest: true

Didn't matter in the test.  I thought it better to test something other than zero values.

File: pkg/authorization/registry/subjectaccessreview/rest.go [hunk 5]
```diff
@@ -112,19 +118,33 @@ func (r *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, err
...
-func (r *REST) isAllowed(ctx kapi.Context, sar *authorizationapi.SubjectAccessReview) error {
-	localSARAttributes := authorizer.DefaultAuthorizationAttributes{
-		Verb:              "create",
-		Resource:          "localsubjectaccessreviews",
-		RequestAttributes: sar,
+func (r *REST) isAllowed(user user.Info, sar *authorizationapi.SubjectAccessReview) error {
+	var localSARAttributes kauthorizer.AttributesRecord
+	// if they are running a personalSAR, create synthentic check for selfSAR
+	if authorizer.IsPersonalAccessReviewFromSAR(sar) {
+		localSARAttributes = kauthorizer.AttributesRecord{
```
Comment (enj) [inline pkg/authorization/registry/subjectaccessreview/rest.go]: This and every other `AttributesRecord` that is manually built is incorrect.  `ResourceRequest` needs to be explicitly set to `true`.

```
I0309 01:03:37.003518   13555 messages.go:126] User "developer" cannot "create" on "" <nil> authorizer.AttributesRecord{User:(*user.DefaultInfo)(0xc4277905c0), Verb:"create", Namespace:"", APIGroup:"authorization.k8s.io", APIVersion:"", Resource:"selfsubjectaccessreviews", Subresource:"", Name:"", ResourceRequest:false, Path:""}
```

Comment (liggitt) [inline pkg/authorization/registry/subjectaccessreview/rest.go]: Well. *That's* a painful default.

File: pkg/authorization/authorizer/interfaces.go [hunk 6]
```diff
@@ -3,47 +3,30 @@ package authorizer
...
-	GetURL() string
-}
-
 // ForbiddenMessageMaker creates a forbidden message from a MessageContext
 type ForbiddenMessageMaker interface {
 	MakeMessage(ctx MessageContext) (string, error)
 }
 
 // MessageContext contains sufficient information to create a forbidden message.  It is bundled in this one object to make it easy and obvious how to build a golang template
 type MessageContext struct {
```
Comment (deads2k) [inline pkg/authorization/authorizer/interfaces.go]: > Drop MessageContext and use authorizer.Attributes directly?

If you want to do it as a followup, sure.  Not needed to collapse interfaces ahead of the rebase.

File: pkg/authorization/registry/subjectaccessreview/rest.go (hunk 5 above)
Comment (deads2k) [inline pkg/authorization/registry/subjectaccessreview/rest.go]: > Well. That's a painful default.

Yeah, guess I missed a couple.

Comment (deads2k) [inline pkg/authorization/registry/subjectaccessreview/rest.go]: Also, glad our tests exercise many of these.  Go us :)

**Comment by deads2k** [comment]
Oh, the cool one was the subtle namespace problem.  We hacked the namespace into the context in a filter for projects.  When I eliminated using context, we lost that value for project authorization which made everything go haywire.

**Comment by deads2k** [comment]
comments addressed.

**Comment by ncdc** [comment]
These look like they might be caused by this PR?

```
09:22:16 === BEGIN TEST CASE ===
09:22:16 test/cmd/images_tests.sh:204: executing 'oc tag cmd-images-old-policy/mysql:5.5 newrepo:latest' expecting success
09:22:16 
FAILURE after 0.722s: test/cmd/images_tests.sh:204: executing 'oc tag cmd-images-old-policy/mysql:5.5 newrepo:latest' expecting success: the command returned the wrong error code
09:22:16 There was no output from the command.
09:22:16 Standard error from the command:
09:22:16 The ImageStream "newrepo" is invalid: []: Internal error: imagestreams "newrepo" is invalid: spec.tags[latest].from: Forbidden: cmd-images-old-policy/mysql
09:22:16 === END TEST CASE ===
09:22:16 In suite "github.com/openshift/origin/test/cmd/policy", test case "test/cmd/policy.sh:113: executing 'oc policy scc-subject-review -f ${OS_ROOT}/test/testdata/job.yaml -o=jsonpath={.status.AllowedBy.name}' expecting success and text 'restricted'" failed:
09:22:16 === BEGIN TEST CASE ===
09:22:16 test/cmd/policy.sh:113: executing 'oc policy scc-subject-review -f ${OS_ROOT}/test/testdata/job.yaml -o=jsonpath={.status.AllowedBy.name}' expecting success and text 'restricted'
09:22:16 FAILURE after 0.194s: test/cmd/policy.sh:113: executing 'oc policy scc-subject-review -f ${OS_ROOT}/test/testdata/job.yaml -o=jsonpath={.status.AllowedBy.name}' expecting success and text 'restricted': the command returned the wrong error code; the output content test failed
09:22:16 There was no output from the command.
09:22:16 Standard error from the command:
09:22:16 error: unable to compute Pod Security Policy Subject Review for "hello": User "bob" cannot create podsecuritypolicyselfsubjectreviews in project "bob"
09:22:16 === END TEST CASE ===
```

**Comment by deads2k** [comment]
> These look like they might be caused by this PR?

Yeah, I'll make it through.

File: pkg/cmd/server/origin/master_config.go [hunk 7]
```diff
@@ -827,8 +827,9 @@ func newAuthorizationAttributeBuilder(requestContextMapper kapi.RequestContextMa
 		requestInfoFactory,
 	)
 	personalSARRequestInfoResolver := authorizer.NewPersonalSARRequestInfoResolver(browserSafeRequestInfoResolver)
+	projectRequestInfoResolver := authorizer.NewProjectRequestInfoResolver(personalSARRequestInfoResolver)
```
Comment (enj) [inline pkg/cmd/server/origin/master_config.go]:
```go
final := chainFunc(
    browserSafeRequestInfoResolver,  // base info
    authorizer.NewPersonalSARRequestInfoResolver,  // wrapper funcs
    authorizer.NewProjectRequestInfoResolver,
)
```

Comment (deads2k) [inline pkg/cmd/server/origin/master_config.go]: > final := chainFunc(
>     browserSafeRequestInfoResolver,  // base info
>     authorizer.NewPersonalSARRequestInfoResolver,  // wrapper funcs
>     authorizer.NewProjectRequestInfoResolver,
> )

No real disagreement, but I suspect this gets reswizzled in the rebase.  You want the related updates now?

**Comment by enj** [comment]
LGTM if tests pass.

File: pkg/cmd/server/origin/master_config.go (hunk 7 above)
Comment (enj) [inline pkg/cmd/server/origin/master_config.go]: After rebase sounds good to me.

**Comment by deads2k** [comment]
re[test]

**Comment by deads2k** [comment]
re[test]

**Comment by deads2k** [comment]
[merge]

**Pull request merged**